lldb/source/Breakpoint/WatchpointResource.cpp

```diff
@@ -0,0 +1,96 @@
+//===-- WatchpointResource.cpp --------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "lldb/Breakpoint/WatchpointResource.h"
+
+using namespace lldb;
+using namespace lldb_private;
+
+WatchpointResource::WatchpointResource(lldb::addr_t addr, size_t size,
+                                       bool read, bool write)
+    : m_id(LLDB_INVALID_WATCHPOINT_RESOURCE_ID), m_addr(addr), m_size(size),
+      m_watch_read(read), m_watch_write(write) {}
+
+WatchpointResource::~WatchpointResource() {
+  std::lock_guard<std::recursive_mutex> guard(m_owners_mutex);
+  m_owners.Clear();
+}
+
+void WatchpointResource::GetMemoryRange(lldb::addr_t &addr,
+                                        size_t &size) const {
+  addr = m_addr;
+  size = m_size;
+}
+
+addr_t WatchpointResource::GetAddress() const { return m_addr; }
+
+size_t WatchpointResource::GetByteSize() const { return m_size; }
+
+void WatchpointResource::GetType(bool &read, bool &write) const {
+  read = m_watch_read;
+  write = m_watch_write;
+}
+
+void WatchpointResource::SetType(bool read, bool write) {
+  m_watch_read = read;
+  m_watch_write = write;
+}
+
+wp_resource_id_t WatchpointResource::GetID() const { return m_id; }
+
+void WatchpointResource::SetID(wp_resource_id_t id) { m_id = id; }
+
+bool WatchpointResource::Contains(addr_t addr) {
+  if (addr >= m_addr && addr < m_addr + m_size)
+    return true;
+  return false;
```

Comment:
`return add >= m_addr && add < m_addr + m_size;`

Reply:
yeah this kind of thing always makes me think 10% harder when I'm reading code, but I know I'm in the minority.